### DISCUSSION TITLE
GraphQL: How to get a list of commits between tags or releases

### DISCUSSION THREAD

**Discussion by josh803316** [community/community#24779] [Apps, API and Webhooks]
<p>I posted on StackOverflow as well: <a href="https://stackoverflow.com/questions/56401516/github-graphql-how-to-get-a-list-of-commits-between-tag" rel="nofollow noopener">github-graphql-how-to-get-a-list-of-commits-between-tag</a></p>
<p>If possible, I would like to make a single query that would give me the commit list between releases or tags. If only 1 tag were provided it would be the list from current master (latest commit) until that tag, and if 2 were provided it would be the list between the 2 tags. I wasn’t sure of the best way to approach this without using more than 1 query. In truth, I only need commits from master to the most recent tag or between tags next to one another if that makes it easier.</p>

**Comment by lee-dohm** [comment]
<p>This kind of query isn’t something that the GraphQL API is designed to be able to give you at this point. In theory, if you’re certain that <code>tag-x</code> is an ancestor of <code>branch-y</code>, and you have the commit hash of <code>tag-x</code>, then you could go from the Commit object pointed to by <code>branch-y</code> and paginate through all commits (using the <a href="https://developer.github.com/v4/object/commit/" rel="nofollow noopener">history connection</a>) until you find the commit hash you’re looking for. You could also use the <code>since</code> parameter on the history connection to prevent you from overwalking. But that doesn’t solve for the case where <code>y</code> is the ancestor of <code>x</code> or that <code>x</code> and <code>y</code> are not proper ancestors of one or the other.</p>
<p>For now, using git itself is probably the best way to get at the information you’re looking for.</p>
<p>I hope that helps!</p>

**Comment by josh803316** [comment]
<p>Thanks <a class="mention" href="/lee-dohm">@lee-dohm</a> for the helpful response! In truth, I’m only ever concerned with the master branch, and reality is that I need to see a commit list from the last release to the current point in time, as a new release is about to be generated.  So even if am just getting a list of commits up to the last tag, or list of commits up to the last release that would be enough. Is that possible with the “Release” or “Tag” and … on Commit fragment? I’m just learning GraphQl (first couple of days) so forgive the ignorance. </p>
<p>Even if there was a way for me to list a bunch of commits, and discern that one of them was tagged, that would be enough, but I can’t seem to get that info just from “Commit” history.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>So if you only want to get the list of commits from the <code>HEAD</code> of the <code>master</code> branch back to some release tag, you could use this query:</p>
<pre><code class="lang-auto">{
  repository(owner: "lee-dohm", name: "octicons-ex") {
    object(expression: "v0.5.0") {
      ... on Commit {
        oid
        messageHeadline
        committedDate
        author {
          user {
            login
          }
        }
      }
    }
  }
}
</code></pre>
<p>to get the information on the commit pointed to by the given tag (“v0.5.0” in my example). From there you could get the <code>committedDate</code> value to supply to this query:</p>
<pre><code class="lang-auto">{
  repository(owner: "lee-dohm", name: "octicons-ex") {
    nameWithOwner
    object(expression: "master") {
      ... on Commit {
        oid
        history(first: 100, since: "2018-06-22T04:37:29Z") {
          nodes {
            oid
            messageHeadline
            author {
              user {
                login
              }
            }
            committedDate
          }
        }
      }
    }
  }
}
</code></pre>
<p>by placing the <code>committedDate</code> in the <code>since</code> parameter (replacing “2018-06-22T04:37:29Z” in my example). That will list all the commits between the tip of the <code>master</code> branch and the given tag.</p>
<p>I hope that helps!</p>

**Comment by josh803316** [comment]
<p><a class="mention" href="/lee-dohm">@lee-dohm</a> This is truly helpful.  I made it a little more dynamic by doing this:</p>
<pre><code>{
  repository(owner: "CoolCompany", name: "awesome-ui") {
    releases(last: 1) {
      edges{
        node{
          tagName
          createdAt
        }
      }
    }
  }
}
</code></pre>
<p>Then I can use the createdAt you suggested in the same way and get my commit list since the last release. I really appreciate the help! It’s too bad I couldn’t do this all in a single query but it took me 3 requests with the v3 rest API so it’s still an improvement and the filtering logic is much easier now. Thank you!</p>

**Comment by lee-dohm** [comment]
<p>You’re welcome :grinning:</p>
<p>One other thing to note is that you don’t need to do <code>edges</code> immediately followed by <code>node</code> in your queries unless you’re getting the <code>pageInfo</code> or <code>totalCount</code> from the <code>edges</code> connection. If you just want the stuff in <code>node</code> (as it appears you do in your query), it can be simplified to:</p>
<pre><code class="lang-auto">{
  repository(owner: "CoolCompany", name: "awesome-ui") {
    releases(last: 1) {
      nodes {
        tagName
        createdAt
      }
    }
  }
}
</code></pre>
<p>I hope that helps!</p>

**Comment by jean** [comment]
<p>I’m curious how I would know what is the previous release. E.g. in one repo I have v1.1.0-beta3, v1.1.1, v1.1.0-beta4.<br>
The github UI knows that the changelog for the first is <code>v1.1.0-beta3..v1.1.0-beta4</code>, but how I do know that from the GraphQL response?<br>
I could do a “human-sort” on the release names, but that wouldn’t be guaranteed to work always.</p>
<p>Here’s the query (roughly):</p>
<pre><code class="lang-auto">...
releases(first: 100) {
      edges {
        node {
          name
          publishedAt
        }
      }
    }
...
</code></pre>
<p>Here’s the result:</p>
<pre><code class="lang-auto">          {
            "node": {
              "name": "v1.1.0-beta.3",
              "publishedAt": "2020-12-15T06:16:07Z"
            }
          },
          {
            "node": {
              "name": "v1.1.1",
              "publishedAt": "2020-12-15T07:03:35Z"
            }
          },
          {
            "node": {
              "name": "v1.1.0-beta.4",
              "publishedAt": "2021-04-07T13:23:00Z"
            }
          },
</code></pre>
